src/js/store/middleware/feature/form.js

```diff
@@ -80,16 +100,34 @@ const handleLoadIndexResponse = (store, next, action) => {
   next(setIndexLoaded({ state: true, feature: FORM }))
 }
 
+const handleFetchSchemaSuccess = (store, next, action) => {
+  next(setLoader({ state: false, feature: FORM }))
+  store.dispatch(setSchema({ schema: action.payload }))
+
+  const state = store.getState()
+  const key = localStorageKey(state)
+
+  let storageState
+  try {
+    const storageStateJSON = window.localStorage.getItem(key)
+    storageState = storageStateJSON && JSON.parse(storageStateJSON)
+  } catch(error) {
+    console.error(error)
+  }
+
+  if (storageState && JSON.stringify(storageState.schema) === JSON.stringify(action.payload)) {
+    store.dispatch(setAskingToRestore({ state: true, feature: FORM }))
+  } else {
+    store.dispatch(rejectLocalStorage())
+  }
```

Comment:
We change our schema loading logic a bit. When we get the form's schema back from the API (identified as an `API_SUCCESS` action that's in response to a `FETCH_SCHEMA` action) we then check local storage for unsubmitted data for the active form.

If we don't find any or if the schema has changed at all in the meantime, we trigger a `REJECT_LOCAL_STORAGE` action, automatically discarding any data that's there without any user input. This is because there either wasn't any data there or it's unusable due to the schema change. This results in the user seeing an empty form populate with default values.

If we _do_ find local data, then we set the `askingToRestore` flag that the `App` component uses to decide whether to prompt the user about local data. That will trigger a prompt and, based on the user's input, either a `REJECT_LOCAL_STORAGE` action or an `ACCEPT_LOCAL_STORAGE` action.

Reply:
Line 118 is the equality check you were talking about in [your comment about how aggressively we're checking for schema consistency](https://github.com/associatedpress/harvester/pull/21#issuecomment-977357382), right? I agree with your take in that comment that erring toward the stricter side is preferable, but I do wonder at some point how we could make a more forgiving comparison. (For now, even a typo fix in the help text of one field would cause this equality check to be false, right?)

Reply:
Exactly! Because we're serializing the schemas in their entirety and checking for string equality, any change to help text or the headline or anything like that would cause stored data to be discarded. That's something I'd like to think a bit more about in the future, but since this takes us from _never_ being able to recover that data to _usually_ being able to recover that data I think this is progress.

Reply:
Yep, definitely progress. This week's other question about deserialization errors for existing data after adding `multiple:true` to a field has been making me wonder more broadly about what it'd look like to diff a Harvester schema.

In particular, I'm thinking through what it'd look like to interpret those diffs as migrations (à la [Rails](https://guides.rubyonrails.org/v6.1/active_record_migrations.html) or [Django](https://docs.djangoproject.com/en/3.2/topics/migrations/)) whether we could add some optional helpers (like an [installable menu](https://developers.google.com/apps-script/guides/menus) in a Google sheet) that could parse recent revisions to the `schema` sheet and suggest updates to the existing `entry` content to match. That's probably overcomplicating things, though.

Reply:
Yeah if the tool could help you out with that kind of thing a bit more that would be nice. I'd be a little hesitant to lean _more_ into Google-Sheets-specific features though since the number one request is generally to use a data store other than Google Sheets, but it would be cool to explore the possibilities!